project/src/main/java/com/google/sps/Models/Database.java

```diff
@@ -128,59 +128,54 @@ public static Document createDocument(String name, String language, String hash,
       docEntity.setProperty("userIDs", userIDs);
       getDatastore().put(docEntity);
 
+      addDocumentForUser(hash, userID);
       return new Document(name, language, hash, userIDs);
   }
 
   public static Document getDocumentByHash(String hash) {
     Query query = new Query("Document").addFilter("hash", Query.FilterOperator.EQUAL, hash);
     Entity docEntity = getDatastore().prepare(query).asSingleEntity();
 
-    String name = (String) docEntity.getProperty("name");
-    String language = (String) docEntity.getProperty("language");
-    ArrayList<Long> userIDs = (ArrayList)docEntity.getProperty("userIDs");
-
     if(docEntity == null) {
       return null;
     }
 
+    String name = (String) docEntity.getProperty("name");
+    String language = (String) docEntity.getProperty("language");
+    ArrayList<Long> userIDs = (ArrayList)docEntity.getProperty("userIDs");
     return new Document(name, language, hash, userIDs);
   }
 
   public static ArrayList<Long> getDocumentUsers(String hash) {
     Query query = new Query("Document").addFilter("hash", Query.FilterOperator.EQUAL, hash);
     Entity docEntity = getDatastore().prepare(query).asSingleEntity();
 
-    ArrayList<Long> userIDs = (ArrayList)docEntity.getProperty("userIDs");
-
     if(docEntity == null) {
       return null;
     }
 
+    ArrayList<Long> userIDs = getListProperty(docEntity, "userIDs");
     return userIDs;
   } 
 
-  public static ArrayList<String> getUsersDocumentsHashes(long userID) {
-    Query query = new Query("User").addFilter("userID", Query.FilterOperator.EQUAL, userID);
-    Entity userEntity = getDatastore().prepare(query).asSingleEntity();
-
-    ArrayList<String> docHashes = (ArrayList)userEntity.getProperty("docHashes");
-
-    if(userEntity == null) {
-      return null;
-    }
-
-    return docHashes;
-  }
-
   public static ArrayList<Document> getUsersDocuments(long userID) {
     ArrayList<String> docHashes = getUsersDocumentsHashes(userID);
-
     ArrayList<Document> docs = new ArrayList<Document>();
     for(String hash : docHashes) {
         Document doc = getDocumentByHash(hash);
         docs.add(doc);
     }
-
     return docs;
   }
+
+  // Datastore does not support empty collections (it will be stored as null)
+  // https://cloud.google.com/appengine/docs/standard/java/datastore/entities#Using_an_empty_list
+  private static ArrayList getListProperty(Entity entity, String prop) {
+    ArrayList items = (ArrayList) entity.getProperty(prop);
```

Comment:
this is because datastore returns null if the arraylist is empty correct?

Reply:
correct. Therefore if the datastore contains null, this will just create an empty list